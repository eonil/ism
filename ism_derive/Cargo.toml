# Note: Generated by AI assistant
[package]
name = "ism_derive"
version.workspace = true
edition.workspace = true
authors.workspace = true
license.workspace = true

[lib]
proc-macro = true

[dependencies]
syn.workspace = true
quote.workspace = true
proc-macro2.workspace = true

[dev-dependencies]
ism = { path = "../ism" }
serde = { workspace = true, features = ["derive"] }
serde_json = { workspace = true } 