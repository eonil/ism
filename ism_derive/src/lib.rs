// Note: Generated by AI assistant
use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, DeriveInput};

#[proc_macro_derive(Serialize)]
pub fn derive_serialize(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    let name = &input.ident;
    
    let expanded = quote! {
        impl ism::Serialize for #name {
            fn serialize(&self) -> ism::text::Message {
                serde_json::to_value(self)
                    .map(|v| serde_json::from_value(v).unwrap())
                    .unwrap()
            }
        }
    };

    TokenStream::from(expanded)
}

#[proc_macro_derive(Deserialize)]
pub fn derive_deserialize(input: TokenStream) -> TokenStream {
    let input = parse_macro_input!(input as DeriveInput);
    let name = &input.ident;

    let expanded = quote! {
        impl ism::Deserialize for #name {
            fn deserialize(msg: ism::text::Message) -> Result<Self, ism::text::Error> {
                let value = serde_json::to_value(msg)?;
                Ok(serde_json::from_value(value)?)
            }
        }
    };

    TokenStream::from(expanded)
} 