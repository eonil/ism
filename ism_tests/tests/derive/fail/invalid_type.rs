// Note: Generated by AI assistant
use ism_derive::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
struct InvalidType {
    num: i32,  // Should fail - numbers not supported
    flag: bool,  // Should fail - booleans not supported
}

fn main() {} 