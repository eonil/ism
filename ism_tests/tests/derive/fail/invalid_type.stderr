error: Only String type is supported in ISM
  --> tests/derive/fail/invalid_type.rs:5:11
   |
5  |     num: i32,
   |          ^^^

error: Only String type is supported in ISM
  --> tests/derive/fail/invalid_type.rs:6:12
   |
6  |     flag: bool,
   |           ^^^^ 